src/main/java/org/mockito/internal/junit/JUnitRule.java

```diff
@@ -8,87 +8,76 @@
+                } finally {
+                    DefaultMockitoFramework.tryRemoveListener(listener);
+                }
 
-    public JUnitRule silent() {
-        return new JUnitRule(logger, true);
-    }
+                //If the 'testFinished' fails below, we don't see the original failure, thrown later
+                DefaultTestFinishedEvent event = new DefaultTestFinishedEvent(target, method.getName(), testFailure);
+                listener.testFinished(event);
 
-    private class SilentStatement extends Statement {
-        private final Object target;
-        private final Statement base;
+                if (testFailure != null) {
+                    throw testFailure;
+                }
 
-        public SilentStatement(Object target, Statement base) {
-            this.target = target;
-            this.base = base;
-        }
+                //Validate only when there is no test failure to avoid reporting multiple problems
+                //This could be part of the listener but to avoid duplication I left it here:
+                Mockito.validateMockitoUsage();
+            }
 
-        public void evaluate() throws Throwable {
-            MockitoAnnotations.initMocks(target);
-            base.evaluate();
-            Mockito.validateMockitoUsage();
-        }
+            private Throwable evaluateSafely(Statement base) {
+                try {
+                    base.evaluate();
+                    return null;
+                } catch (Throwable throwable) {
+                    return throwable;
+                }
+            }
+        };
     }
 
-    private class DefaultStatement extends Statement {
-        private final Object target;
-        private final String testName;
-        private final Statement base;
-
-        DefaultStatement(Object target, String testName, Statement base) {
-            this.target = target;
-            this.testName = testName;
-            this.base = base;
-        }
-
-        public void evaluate() throws Throwable {
-            RuleStubbingHintsReporter reporter = new RuleStubbingHintsReporter();
-            Mockito.framework().addListener(reporter);
-            try {
-                performEvaluation(reporter);
-            } finally {
-                Mockito.framework().removeListener(reporter);
-            }
-        }
+    public JUnitRule silent() {
+        return new JUnitRule(logger, Strictness.SILENT);
+    }
 
-        private void performEvaluation(RuleStubbingHintsReporter reporter) throws Throwable {
-            MockitoAnnotations.initMocks(target);
-            try {
-                base.evaluate();
-            } catch(Throwable t) {
-                reporter.getStubbingArgMismatches().format(testName, logger);
-                throw t;
-            }
-            reporter.getUnusedStubbings().format(testName, logger);
-            Mockito.validateMockitoUsage();
-        }
+    public JUnitRule strictStubs() {
```

Comment:
Great questions and feedback.

silent() and strictStubs() are different: the first silences all warnings whatever they are (currently , the only warnings are related to stubbing but it will change in the future). The latter specifically makes the stubbing stricter. It does not do Mockito stricter, for example, it does not make the mock objects "strict".

Thoughts/ideas/suggestions for improvements?